lib/src/rules/library_annotations.dart

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2022, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+// ignore: implementation_imports
+import 'package:analyzer/src/dart/element/extensions.dart';
+import 'package:meta/meta_meta.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Attach library annotations to library directives.';
+
+const _details = r'''
+Attach library annotations to library directives, rather than
+some other library-level element.
+
+**BAD:**
+```dart
+import 'package:test/test.dart';
+
+@TestOn('browser')
+void main() {}
+```
+
+**GOOD:**
+```dart
+@TestOn('browser')
+library;
+import 'package:test/test.dart';
+
+void main() {}
+```
+
+**NOTE:** An unnamed library, like `library;` above, is only supported in Dart
+2.19 and later. Code which might run in earlier versions of Dart will need to
+provide a name in the `library` directive.
+''';
+
+class LibraryAnnotations extends LintRule {
+  static const LintCode code = LintCode('library_annotations',
+      'This annotation must be attached to a library directive.',
+      correctionMessage: 'Attach library annotations to library directives.');
+
+  LibraryAnnotations()
+      : super(
+            name: 'library_annotations',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  LintCode get lintCode => code;
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    var visitor = _Visitor(this);
+    registry.addClassDeclaration(this, visitor);
+    registry.addClassTypeAlias(this, visitor);
+    registry.addEnumDeclaration(this, visitor);
+    registry.addExportDirective(this, visitor);
+    registry.addExtensionDeclaration(this, visitor);
+    registry.addFunctionTypeAlias(this, visitor);
+    registry.addFunctionDeclaration(this, visitor);
+    registry.addGenericTypeAlias(this, visitor);
+    registry.addImportDirective(this, visitor);
+    registry.addMixinDeclaration(this, visitor);
+    registry.addTopLevelVariableDeclaration(this, visitor);
```

Comment:
Consider visiting the `CompilationUnit` instead. I believe that there will be other diagnostics if the annotation isn't on the first member of the unit (directive or declaration), so we should only need to check the first member's annotations.

We might also want to bail if the unit is a part because there should also be diagnostics produced elsewhere in those cases.

Reply:
Done, and I agree about the part file. Done.